src/test/java/com/microsoft/azure/relay/HybridConnectionListenerTest.java

```diff
@@ -29,45 +33,135 @@
+			offlineHandlerCalled.set(true);
+		});
+		
+		listener.openAsync(Duration.ofSeconds(15)).join();
+		assertTrue("Listener failed to open.", listener.isOnline());
+		assertTrue("Listener open handler was not called", onlineHandlerCalled.get());
+		
+		listener.close();
+		assertFalse("Listner should be closed", listener.isOnline());
+		assertTrue("Listener offline handler was not called", offlineHandlerCalled.get());
 	}
 	
 	@Test
-	public void openAsyncTest() {
-		assertTrue("Listener failed to open.", listener.isOnline());
+	public void webSocketConnectionTest() throws InterruptedException, ExecutionException, TimeoutException {
+		int waitMS = 500;
+		HybridConnectionListener listener = new HybridConnectionListener(CONNECTION_URI, tokenProvider);
+		listener.openAsync(Duration.ofSeconds(15)).join();
+		
+		// Client initiates closing
+		CompletableFuture<Void> clientClosing = listener.acceptConnectionAsync().thenCombineAsync(client.createConnectionAsync(), (rendezvousConnection, clientConnection) -> {
+			assertTrue("client connection should be open", clientConnection != null && clientConnection.isOpen());
+			assertTrue("rendezvous connection should be open", rendezvousConnection != null && rendezvousConnection.isOpen());
+			clientConnection.closeAsync().join();
+			
+			assertTrue("listener should still be open", listener.isOnline());
+			assertFalse("client connection should be closed", clientConnection.isOpen());
+			return rendezvousConnection;
+		}).thenCompose(rendezvousConnection -> {
+			return CompletableFutureUtil.delayAsync(Duration.ofMillis(waitMS), EXECUTOR).whenComplete(($void, ex) -> {
+				assertFalse("rendezvous connection should be closed implicitly when client connection closes", rendezvousConnection.isOpen());
+			});
+		});
+		
+		// Rendezvous initiates closing
+		CompletableFuture<Void> rendezvousClosing = client.createConnectionAsync().thenCombineAsync(listener.acceptConnectionAsync(), (rendezvousConnection, clientConnection) -> {
+			assertTrue("client connection should be open", clientConnection != null && clientConnection.isOpen());
+			assertTrue("rendezvous connection should be open", rendezvousConnection != null && rendezvousConnection.isOpen());
+			rendezvousConnection.closeAsync().join();
+			
+			assertTrue("listener should still be open", listener.isOnline());
+			assertFalse("rendezvous connection should be closed", rendezvousConnection.isOpen());
+			return clientConnection;
+		}).thenCompose(clientConnection -> {
+			return CompletableFutureUtil.delayAsync(Duration.ofMillis(waitMS), EXECUTOR).whenComplete(($void, ex) -> {
+				assertFalse("client connection should be closed implicitly when rendezvous connection closes", clientConnection.isOpen());
+			});
+		});
+		
+		CompletableFuture.allOf(clientClosing, rendezvousClosing).whenComplete(($void, ex) -> {
+			if (ex != null) {
+				fail(ex.getMessage());
+			}	
+			listener.close();
```

Comment:
>listener.close(); [](start = 3, length = 17)

I'd strongly recommend putting the listener close inside a try/finally block. If a test case fails but leaves a listener connected to the same endpoint it could easily mess with other tests.